src/gss.c

```diff
@@ -0,0 +1,607 @@
+			pthread_cond_wait(&gss_data_available, &gss_queue_tail_mutex);
+		}
+
+		pthread_mutex_unlock(&gss_queue_tail_mutex);
+
+		log_debug("gss_auth_worker(): processing slot %d", current_slot);
+
+		/* We have at least one request in the queue */
+		request = &gss_auth_queue[current_slot];
+		current_slot = (current_slot + 1) % GSS_REQUEST_QUEUE_SIZE;
+
+		if (gss_recvauth(request)) {
+			request_status = GSS_STATUS_SUCCESS;
+		} else {
+			request_status = GSS_STATUS_FAILED;
+		}
+		set_request_status(request, request_status);
+
+		log_debug("gss_auth_worker(): authentication completed, status=%d", request_status);
+	}
+
+	return NULL;
+}
+
+
+static bool gss_recvauth(struct gss_auth_request *request)
+{
+	gss_buffer_desc send_tok;
+	OM_uint32 maj_stat,
+		  min_stat,
+		  lmin_s;
+	gss_buffer_desc gbuf;
+	gss_cred_id_t server_credentials, delegated_creds;
+
+	if (!initialize_gss_parameters(request, request->client->gss_parameters)) {
+		return false;
+	}
+
+	if (!cf_auth_krb_server_keyfile) {
+		log_debug("No cf_auth_krb_server_keyfile specified in config");
+		return false;
+	}
+	setenv("KRB5_KTNAME", cf_auth_krb_server_keyfile, 1);
+	if (GSS_INITIAL == request->gss_state) {
+		request->context = GSS_C_NO_CONTEXT;
+	}
+
+	gbuf.length = request->token_length;
+	gbuf.value = request->token;
+
+	log_debug("Received GSSAPI token.");
+	server_credentials = GSS_C_NO_CREDENTIAL;
+	delegated_creds = GSS_C_NO_CREDENTIAL;
+
+	/* Accept the context. */
+	maj_stat = gss_accept_sec_context(&min_stat,
+					  &request->context,
+					  server_credentials,
+					  &gbuf,
+					  GSS_C_NO_CHANNEL_BINDINGS,
+					  &request->gss_name,
+					  NULL,
+					  &send_tok,
+					  &request->flags,
+					  NULL,
+					  cf_auth_gss_accept_delegation ? &delegated_creds : NULL);
+
+	log_debug("gss_accept_sec_context major: %u, "
+		  "minor: %u, outlen: %u, outflags: %x",
+		  maj_stat, min_stat,
+		  (unsigned int) send_tok.length, request->flags);
+
+	/* Send back token to the client, if expected to do so. */
+	if (send_tok.length != 0) {
+		PktBuf _buf;
+		uint8_t _data[512];
+
+		log_debug("sending GSS response token of length %u", (unsigned int) send_tok.length);
+
+		/* Construct a custom response with the token */
+		pktbuf_static(&_buf, _data, sizeof(_data));
+		pktbuf_put_char(&_buf, 'R');
+		pktbuf_put_uint32(&_buf, send_tok.length + 4 + 4);
+		pktbuf_put_uint32(&_buf, AUTH_REQ_GSS_CONT);
+		pktbuf_put_bytes(&_buf, send_tok.value, send_tok.length);
+
+		if (!pktbuf_send_immediate(&_buf, request->client)) {
+			gss_release_buffer(&lmin_s, &send_tok);
+			log_debug("Failed to send gss token to client");
+			return false;
+		}
+		gss_release_buffer(&lmin_s, &send_tok);
+	}
+
+	if (maj_stat != GSS_S_COMPLETE && maj_stat != GSS_S_CONTINUE_NEEDED) {
+		gss_delete_sec_context(&lmin_s, &request->context, GSS_C_NO_BUFFER);
+		log_debug("accepting GSS security context failed major: %u, minor: %u",
+			  maj_stat, min_stat);
+		return false;
```

Comment:
The same necessary checking for calling `gss_release_cred(&min_stat, &server_credentials);` if the return happens. Or if this is not met, then there is no need to call `gss_release_cred(&min_stat, &server_credentials);` ?
I am not 100% sure for that.